src/engines/hugin-inference-engine.ts

```diff
@@ -0,0 +1,226 @@
+    return found !== []
+  }
+
+  getParents = (name: string) => {
+    if (!this._network[name]) return []
+    return [...this._network[name].parents]
+  }
+
+  // NOTE: in order to prevent external mutation of the CPT, the
+  // result is cloned prior to returning it.
+  getDistribution = (name: string) => clone(this._network[name].cpt)
+
+  // NOTE: in order to prevent external mutation of the CPT, the
+  // input is cloned prior to assigining to the internally cached value.
+  setDistribution = (name: string, cpt: ICptWithParents | ICptWithoutParents) => {
+    const node: INode = this._network[name]
+    const expectedLevels = node.states
+    const expectedParents = node.parents
+
+    // Perform some sanity checks.
+    const err = (reason: string) => {
+      throw new Error(`Cannot set the distribution for ${name}.  ${reason}.`)
+    }
+    if (!node) err('The variable does not exist in the network')
+
+    const observedLevels = Array.isArray(cpt)
+      ? cpt.length > 0
+        ? Object.keys(cpt[0].then)
+        : []
+      : Object.keys(cpt).sort()
+
+    const observedParents = Array.isArray(cpt)
+      ? cpt.length > 0
+        ? Object.keys(cpt[0].when)
+        : []
+      : []
+
+    const hasCorrectLevels: boolean = observedLevels.every(x => expectedLevels.includes(x)) && expectedLevels.every(x => observedLevels.includes(x))
+    const hasCorrectParents: boolean = observedParents.every(x => expectedParents.includes(x)) && observedParents.every(x => expectedParents.includes(x))
+    if (!hasCorrectLevels) err('The provided distribution did not have the expected levels')
+    if (!hasCorrectParents) err('The provided distribution did not have the expected parents')
+
+    // Passed the sanity checks.   reset the cache and set the (cloned) value.
+    this.resetCache()
+    this._network[name].cpt = clone(cpt)
+  }
+
+  // Get the clique potentials.   If the internal cache is not empty, then
+  // return the cached values, otherwise, compute them with the current
+  // evidence
+  private getCliquesPotentials: () => ICliquePotentials = () => {
+    if (Object.keys(this._potentials).length === 0) {
+      const initialPotentials = createInitialPotentials(this._cliques, this._network, this._evidence)
+      const propagatedPotentials = propagatePotentials(this._network, this._junctionTree, this._cliques, this._sepSets, initialPotentials, this._connectedComponents.map(x => x[0]))
+      this._potentials = normalizeCliquePotentials(propagatedPotentials)
+    }
+    return this._potentials
+  }
+
+  hasEvidenceFor = (name: string) => this._evidence[name] != null;
+
+  setEvidence = (evidence: { [name: string]: string }) => {
+    this.removeAllEvidence()
+    this.updateEvidence(evidence)
+  }
+
+  updateEvidence = (evidence: { [name: string]: string }) => {
+    for (const [name, value] of Object.entries(evidence)) {
+      const oldValue = this._evidence[name]
+      if (oldValue == null || oldValue !== value) {
```

Comment:
Now, I realize that you are using `==` only for `null`... but I would always use `===` to keep the pattern and also because you can get used to it and use `==` where you should not use.

```suggestion
      if (oldValue === null || oldValue !== value) {
```